solutions/task_7.py

```diff
@@ -0,0 +1,213 @@
+
+'MixInBase':
+    Class which can be inherited to get ability to mix in classes. (To compare with 'MixInMeta'.)
+
+'DocStyleMeta':
+    Metaclass which can be used to check doc-strings of methods in a class.
+
+'CacheMeta':
+    Metaclass which can be used to cache all methods in a class.
+"""
+import re
+import time
+import unittest
+import types
+from functools import wraps
+
+
+class MixInBase(object):
+
+    @classmethod
+    def mix_in(cls, new_base):
+        cls.__bases__ += (new_base,)
+
+
+class MixInMeta(type):
+
+    def mix_in(cls, new_base):
+        cls.__bases__ += (new_base,)
+
+
+class DocStyleMeta(type):
+
+    def __new__(mcs, name, bases, dct):
+        for key, value in dct.items():
+            if isinstance(value, types.FunctionType) and re.search('__', key) is None:
+                if value.__doc__ is None:
+                    raise DocStyleError("'{}' has no documentation string.".format(key))
+                if re.search('  ', value.__doc__) is not None:
+                    raise DocStyleError("'{}' has more than one space delimiter between words in documentation "
+                                        "string.".format(key))
+                if re.search('[A-Z]', str(key)) is not None:
+                    raise DocStyleError("'{}' has no snake-case style.".format(key))
+        return super().__new__(mcs, name, bases, dct)
+
+
+CACHE = {}  # Cache imitation.
+CACHE_LIFE_TIME = 1  # Cache expiration time.
+
+
+class Creature(object):  # Common parent class for "Man" and "Women", just to not duplicate "sound" method.
+
+    def __init__(self, genus):
+        self.genus = genus
+
+    def sound(self, msg):
+        return "{}: {}".format(self.genus, msg)
+
+
+class Man(Creature, metaclass=MixInMeta):  # Metaclass is used to add "mix in" functionality.
+
+    GENUS = 'Man'
+
+    def __init__(self):
+        super(Man, self).__init__(self.GENUS)
+
+
+class Women(MixInBase, Creature):  # Inheritance is used to add "mix in" functionality.
+
+    GENUS = 'Women'
+
+    def __init__(self):
+        super(Women, self).__init__(self.GENUS)
+
+
+class SingMixin(object):  # Class that will be mixed in during tests.
+
+    def sing(self, sound):
+        return self.sound(sound)
+
+
+class DocStyleError(Exception):  # Common error for all exceptions in "DocStyleMeta" class.
+    pass
+
+
+def cached(cache_life_time):  # Just your caching function from #task_3 to use in "CacheMeta".
+    def wrapper(func):
+
+        CACHE[func] = {}
+
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+            key = '|'.join([str(args), str(kwargs)])
+            result, created_at = CACHE[func].get(key, (None, None))
+            if created_at is None or (time.time() - created_at) >= cache_life_time:
+                result = func(*args, **kwargs)
+                CACHE[func][key] = [result, time.time()]
+            return result
+        return wrapped
+
+    return wrapper
+
+
+class CacheMeta(type):
```

Comment:
Why did you place this class here? (Give me the proof of the reason)

Reply:
Because cached function is a wrapper creator and when I call it with some argument it returns me a wrapper which I use for wrapping of a value. Because wrappers normally shall be defined before using them for wrapping of some function CacheMeta class is placed here. Probably it works like this but I'm not sure.